clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java

```diff
@@ -169,8 +150,20 @@ public DeploymentResult operate(List priorOutputs) {
     return makeDeploymentResult(service);
   }
 
+  private EcsServerGroupName buildEcsServerGroupName(AmazonECS ecs, Namer<EcsResource> namer) {
+    EcsServerGroupNameResolver serverGroupNameResolver =
+        new EcsServerGroupNameResolver(description.getEcsClusterName(), ecs, getRegion(), namer);
+
+    if (description.getMoniker() != null) {
```

Comment:
this seems to be resolving the next SG name based on the presence of a `Moniker` in the description vs. the account-level naming strategy. Is it possible (esp if the request is hitting `gate` directly vs. via `deck`) a moniker could be present but differ from the app/stack/detail settings also provided in the description (in which case we would be naming this incorrectly)?

Reply:
Good point. There's nothing validating that the values must match between the current name fields and the `Moniker` fields. We could update the `EcsCreateServerGroupDescriptionValidator` to verify that the values match if they're set and reject the request.

Reply:
Though one thing that helps here is that the `EcsTagNamer` will first parse the service name using the Frigga naming convention, and then overwrite the values for stack, detail, sequence etc with values from the tags if present.

Reply:
So without validation the scenario is something like a `CreateServerGroupDescription` with values like

```json
{
  "application": "foo",
  "moniker": {
    "app": "bar"
  }
}
```

| Naming Strategy | Current Service Name | New Service Name |
| - | - | - |
| `default` | | `bar-v001` |
| `tags` | | `bar-v001` |
| `default` | `foo-v001` | `bar-v001` |
| `tags` | `foo-v001` | `bar-v001` |
| `default` | `bar-v001` | `bar-v002` |
| `tags` | `bar-v001` | `bar-v002` |

So I think adding validation would be the best approach as it would prevent a mismatch in outcomes.

Reply:
Would it make more sense to have validation that says both properties cannot be defined? Rather than validating that they're the same?

Reply:
@deverton validating that they match is good, though would it matter as much if we were sure which was the correct strategy to use? Is the account's naming strategy available here and could we use it to make a decision?

@piradeepk I believe deck will populate `moniker` values by default when you specify app/stack/detail so enforcing the presence of only one or the other would likely break a bunch of existing pipelines.

Reply:
Maybe I'm misunderstanding, but should the source of the data change the naming strategy though? As in, if values are set in the `moniker` object, that doesn't mean the code should use tags. And vice versa, just because tags are used as the source of truth for the name, doesn't mean the deployment request needs to use the `moniker` object. The `application`, `freeFormDetails`, and `stack` fields still work.

Reply:
> just because tags are used as the source of truth for the name, doesn't mean the deployment request needs to use the moniker object. The application, freeFormDetails, and stack fields still work.

ok, I did not fully understand this. So if the idea is that these fields should match (enforced through your new validation) and be correctly set during the deploy operation based on lookup of the existing services, then I am not concerned. Your chart of the diff name results based on mismatched input is exactly what I was trying to avoid and it seems we are now protected against it. Thanks for the additional context!

Reply:
Hm, that suggests maybe this PR should be split in to two. One part to handle the addition of tag based name storage, and one to change the deployment request object to prefer the `moniker` structure over the top-level `application`, `freeFormDetails`, and `stack` fields.

Reply:
I'm not likely to get time to address this until next week at the earliest. Would you rather I split the patch up or are you happy to accept this as is?

Reply:
Thanks @deverton for all your hard work! Based on the limited time available, we'd like to see this feature go into the next release, as many in the community would benefit from this! 😃  I've added the merge label